src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -66,23 +76,67 @@ public async Task<BundleResponse> Handle(BundleRequest bundleRequest, Cancellati
         {
             var bundleResource = bundleRequest.Bundle.ToPoco<Hl7.Fhir.Model.Bundle>();
 
-            if (bundleResource.Type != Hl7.Fhir.Model.Bundle.BundleType.Batch)
+            await FillRequestLists(bundleResource.Entry);
+
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Batch)
             {
-                throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.OnlyCertainBundleTypesSupported);
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
+                };
+
+                await ExecuteAllRequests(responseBundle);
+                return new BundleResponse(responseBundle.ToResourceElement());
             }
+            else if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Transaction)
+            {
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.TransactionResponse,
+                };
 
-            await FillRequestLists(bundleResource.Entry);
+                return await ExecuteTransactionForAllRequests(responseBundle);
+            }
 
-            var responseBundle = new Hl7.Fhir.Model.Bundle
+            throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.OnlyCertainBundleTypesSupported);
+        }
+
+        private async Task<BundleResponse> ExecuteTransactionForAllRequests(Hl7.Fhir.Model.Bundle responseBundle)
+        {
+            try
             {
-                Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
-            };
+                using (var transaction = _transactionHandler.BeginTransaction())
+                {
+                    await ExecuteAllRequests(responseBundle);
 
-            await ExecuteAllRequests(responseBundle);
+                    transaction.Complete();
+                }
+            }
+            catch (TransactionAbortedException)
+            {
+                _logger.LogError("Failed to commit a transaction. Throwing BadRequest as a default exception.");
+                throw new BadRequestException("Transaction aborted while processing request");
+            }
 
             return new BundleResponse(responseBundle.ToResourceElement());
         }
 
+        private static void ThrowTransactionException(EntryComponent responseBundle)
+        {
+            int responseStatus = int.Parse(responseBundle?.Response?.Status);
+            string errorMessage = $"Transaction aborted while processing request url: '{responseBundle?.FullUrl}'. {responseBundle?.Response?.Outcome}";
```

Comment:
What does this error message end up looking like to the client? Is it an operation outcome with a message of some string followed by an encoded version of the internal operation outcome? We may be able to instead just return multiple issues in the operation outcome.